hazelcast/src/main/java/com/hazelcast/collection/impl/queue/QueueContainer.java

```diff
@@ -1102,37 +1107,33 @@ void setId(long itemId) {
      * @param mergePolicy  the {@link SplitBrainMergePolicy} instance to apply
      * @return the used {@link QueueItem} if merge is applied, otherwise {@code null}
      */
-    public QueueItem merge(QueueMergeTypes mergingValue, SplitBrainMergePolicy<Data, QueueMergeTypes> mergePolicy) {
+    public Queue<QueueItem> merge(QueueMergeTypes mergingValue,
+                                  SplitBrainMergePolicy<Collection<Object>, QueueMergeTypes> mergePolicy) {
         SerializationService serializationService = nodeEngine.getSerializationService();
         serializationService.getManagedContext().initialize(mergingValue);
         serializationService.getManagedContext().initialize(mergePolicy);
 
-        // try to find an existing item with the same value
-        QueueItem existingItem = null;
-        for (QueueItem item : getItemQueue()) {
-            if (mergingValue.getValue().equals(item.getData())) {
-                existingItem = item;
-                break;
-            }
+        Queue<QueueItem> existingItems = getItemQueue();
+
+        QueueMergeTypes existingValue = createMergingValue(serializationService, existingItems);
+        Collection<Object> newValues = mergePolicy.merge(mergingValue, existingValue);
+
+        if (existingValue == null) {
+            createNewQueueItems(existingItems, newValues, serializationService);
+        } else if (newValues == null || newValues.isEmpty()) {
+            existingItems.clear();
```

Comment:
We can destroy the container from within (there is a reference to the `QueueService`).

It was also wrong to directly call `clear()` and `add()` on `existingItems`. The transaction data is one issue, the `QueueStore` another (I didn't see it before).

The `QueueStore` is a problem on its own. It uses the item ID from the `QueueItem` as key, so the consistency of the store already breaks during the split-brain (duplicate IDs). I tried some things, but since the item IDs are not unique anymore, I found no way to fix the `QueueStore` content properly. I think it's not possible. So I gave up on deleting items, I just add new merged items by using `QueueContainer.offer()` for new values. The `QueueSplitBrainTest` was adapted accordingly.